src/library/compiler/llvm.cpp

```diff
@@ -0,0 +1,1242 @@
+			  lean_string_cstr(tripleStr),
+			  lean_string_cstr(cpuStr),
+			  lean_string_cstr(featuresStr),
+			  optLevel,
+			  relocMode,
+			    codeModel);
+  
+  if (LLVM_DEBUG) {
+    fprintf(stderr, "...%s ; out: %p \n", __PRETTY_FUNCTION__, tm);
+  }
+
+  return lean_io_result_mk_ok(TargetMachine_to_lean(tm));
+}
+
+// opaque getTargetFromTriple (triple: @&String): IO (Ptr Target)
+extern "C" LEAN_EXPORT lean_object *
+lean_llvm_get_target_from_triple(lean_object *triple,
+                     lean_object * /* w */) {
+    if (LLVM_DEBUG) {
+    fprintf(stderr, "%s ; triple: %p \n", __PRETTY_FUNCTION__,
+	    lean_string_cstr(triple));
+    }
+    LLVMTargetRef t;
+    char *errmsg = NULL;
+    int is_error = LLVMGetTargetFromTriple(lean_string_cstr(triple), &t, &errmsg);
+    assert(!is_error && "failed to get target from triple");
+    if (LLVM_DEBUG) {
+      fprintf(stderr, "...%s ; error?: %d \n", __PRETTY_FUNCTION__, is_error);
+      fprintf(stderr, "...%p ; t: %p \n", __PRETTY_FUNCTION__, t);
+    }
+    return lean_io_result_mk_ok(Target_to_lean(t));
+}
+
+// opaque getDefaultTargetTriple: IO String
+extern "C" LEAN_EXPORT lean_object *
+lean_llvm_get_default_target_triple(lean_object* /* w */) {
+  char *triple = LLVMGetDefaultTargetTriple();
+    if (LLVM_DEBUG) {
+      fprintf(stderr, "%s; triple: %s \n", __PRETTY_FUNCTION__, triple);
+    }
+    return lean_io_result_mk_ok(lean::mk_string(triple));
+  
+}
+
+
+// opaque targetMachineEmitToFile (targetMachine: @&Ptr TargetMachine) (module: @&Ptr Module) (filepath: @&String) (codegenType: @&UInt64): IO Unit
+extern "C" LEAN_EXPORT lean_object *
+lean_llvm_target_machine_emit_to_file(lean_object *target_machine,
+				      lean_object *module,
+				      lean_object *filepath,
+				      uint64_t codegenType,
+				      lean_object* /* w */) {
+  // TODO (bollu): move this to a different function
+  LLVMInitializeAllTargetInfos();
+  LLVMInitializeAllTargets();
+  LLVMInitializeAllTargetMCs();
+  LLVMInitializeAllAsmParsers();
+  LLVMInitializeAllAsmPrinters();
+  
+    if (LLVM_DEBUG) {
+      fprintf(stderr, "%s ; target_machine: %p \n", __PRETTY_FUNCTION__, lean_to_TargetMachine(target_machine));
+      fprintf(stderr, "...%s ; module: %p \n", __PRETTY_FUNCTION__, lean_to_Module(module));
+      fprintf(stderr, "...%s ; filepath: %s \n", __PRETTY_FUNCTION__, lean_string_cstr(filepath));
+      fprintf(stderr, "...%s ; codegenType: %lu \n", __PRETTY_FUNCTION__, codegenType);
+    }
+    char *err_msg = NULL;
+    char *filepath_c_str = strdup(lean_string_cstr(filepath)); // TODO(bollu): do we need the `strdup`?
+    // LLVMTargetMachineEmitToFile(LLVMTargetMachineRef T, LLVMModuleRef M, char *Filename, LLVMCodeGenFileType codegen, char **ErrorMessage)
+    int is_error = LLVMTargetMachineEmitToFile(lean_to_TargetMachine(target_machine),
+				lean_to_Module(module),
+				filepath_c_str,
+				LLVMCodeGenFileType(codegenType),
+				&err_msg);
+
```

Comment:
Too many newlines?

Reply:
Resolved by running `clang-format` on the file.